### ISSUE TITLE
How to get the coordinates of bbox when inference stream data?

### ISSUE THREAD

**Issue by hshs0123** [ultralytics/yolov5#4994]
## ❔Question
How to get the coordinates of bbox when inference stream video?
And Is it  faster using detect.py to do the inference compared with pytorch.hub.load().
For me, It seems that using detect.py and pytorch.hub.load() for the same image, detect.py has 140fps, and pytorch.hub.load() has 50 fps.

**Comment by glenn-jocher** [comment]
@hshs0123 detect.py and hub inference should have similar speeds. If you have a reproducible issue demonstrating otherwise please let us know providing a minimum reproducible example.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by hshs0123** [comment]
```model = torch.hub.load('ultralytics/yolov5', 'yolov5s')
last_time = time.time()
results = model("runs/detect/exp/image0.jpg", size=640)
print("fps: {}".format(1 / (time.time() - last_time)))
```
The output is:

E:\anaconda3\envs\pytorch\python.exe C:/Users/10980/PycharmProjects/msstestcapture.py
Using cache found in C:\Users\10980/.cache\torch\hub\ultralytics_yolov5_master
YOLOv5  2021-9-24 torch 1.9.1 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 11264.0MB)

Fusing layers... 
Model Summary: 224 layers, 7266973 parameters, 0 gradients
Adding AutoShape... 
fps: 32.005616219887216


But detect.py gives this result for the same image.

YOLOv5  2021-9-21 torch 1.9.1 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 11264.0MB)

Fusing layers... 
Model Summary: 224 layers, 7056607 parameters, 0 gradients, 16.3 GFLOPs
image 1/1 C:\Users\10980\PycharmProjects\runs\detect\exp\image0.jpg: 640x640 Done. (0.011s)
Speed: 0.0ms pre-process, 11.0ms inference, 0.0ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs\detect\exp18

Process finished with exit code 0

**Comment by hshs0123** [comment]
And how to get the coordinates of bbox when inference stream data? I know there's a save.txt option there. But I want to get the coordinates once I get results. Can I use hub inference to do this?

**Comment by glenn-jocher** [comment]
@hshs0123 your profiling code is not correct as it is not synchronizing CUDA. YOLOv5 Hub models include built-in profiling using correct timing:

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # or yolov5m, yolov5l, yolov5x, custom

# Images
img = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, PIL, OpenCV, numpy, list

# Inference
results = model(img)

# Results
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.

# YOLOv5 🚀 2021-9-28 torch 1.9.1 CPU
# Fusing layers... 
# Model Summary: 224 layers, 7266973 parameters, 0 gradients
# Adding AutoShape... 
# image 1/1: 720x1280 2 persons, 2 ties
# Speed: 753.5ms pre-process, 155.4ms inference, 2.0ms NMS per image at shape (1, 3, 384, 640)
```

See PyTorch Hub tutorial to get started.

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by hshs0123** [comment]
> @hshs0123 your profiling code is not correct as it is not synchronizing CUDA. YOLOv5 Hub models include built-in profiling using correct timing:
> 
> ```python
> import torch
> 
> # Model
> model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # or yolov5m, yolov5l, yolov5x, custom
> 
> # Images
> img = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, PIL, OpenCV, numpy, list
> 
> # Inference
> results = model(img)
> 
> # Results
> results.print()  # or .show(), .save(), .crop(), .pandas(), etc.
> 
> # YOLOv5 🚀 2021-9-28 torch 1.9.1 CPU
> # Fusing layers... 
> # Model Summary: 224 layers, 7266973 parameters, 0 gradients
> # Adding AutoShape... 
> # image 1/1: 720x1280 2 persons, 2 ties
> # Speed: 753.5ms pre-process, 155.4ms inference, 2.0ms NMS per image at shape (1, 3, 384, 640)
> ```
> 
> See PyTorch Hub tutorial to get started.
> 
> ## YOLOv5 Tutorials
> * [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)  🚀 RECOMMENDED
> * [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)  ☘️ RECOMMENDED
> * [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)  🌟 NEW
> * [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)  🌟 NEW
> * [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
> * [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)  ⭐ NEW
> * [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
> * [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
> * [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
> * [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
> * [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
> * [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)  ⭐ NEW
> * [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

okay, thank you! How to improve the inference speed then? And my main question is how to get the coordinates of bbox when inference stream data? I know there's a save.txt option there. But I want to get the coordinates once I get results. Can I use hub inference to do this?

**Comment by hshs0123** [comment]
> @hshs0123 your profiling code is not correct as it is not synchronizing CUDA. YOLOv5 Hub models include built-in profiling using correct timing:
> 
> ```python
> import torch
> 
> # Model
> model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # or yolov5m, yolov5l, yolov5x, custom
> 
> # Images
> img = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, PIL, OpenCV, numpy, list
> 
> # Inference
> results = model(img)
> 
> # Results
> results.print()  # or .show(), .save(), .crop(), .pandas(), etc.
> 
> # YOLOv5 🚀 2021-9-28 torch 1.9.1 CPU
> # Fusing layers... 
> # Model Summary: 224 layers, 7266973 parameters, 0 gradients
> # Adding AutoShape... 
> # image 1/1: 720x1280 2 persons, 2 ties
> # Speed: 753.5ms pre-process, 155.4ms inference, 2.0ms NMS per image at shape (1, 3, 384, 640)
> ```
> 
> See PyTorch Hub tutorial to get started.
> 
> ## YOLOv5 Tutorials
> * [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)  🚀 RECOMMENDED
> * [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)  ☘️ RECOMMENDED
> * [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)  🌟 NEW
> * [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)  🌟 NEW
> * [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
> * [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)  ⭐ NEW
> * [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
> * [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
> * [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
> * [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
> * [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
> * [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)  ⭐ NEW
> * [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

E:\anaconda3\envs\pytorch\python.exe C:/yolov5-master/detect.py
detect: weights=C:/Users/10980/PycharmProjects/best.pt, source=runs/detect/exp25/zidane.jpg, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=0, view_img=True, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False
YOLOv5  2021-9-21 torch 1.9.1 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 11264.0MB)

Fusing layers... 
Model Summary: 224 layers, 7056607 parameters, 0 gradients, 16.3 GFLOPs
image 1/1 C:\yolov5-master\runs\detect\exp25\zidane.jpg: 384x640 2 person_heads, 2 person_vboxs, Done. (0.016s)
Speed: 0.0ms pre-process, 15.6ms inference, 0.0ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs\detect\exp26
#############################################

E:\anaconda3\envs\pytorch\python.exe C:/Users/10980/PycharmProjects/msstestcapture.py
Using cache found in C:\Users\10980/.cache\torch\hub\ultralytics_yolov5_master
YOLOv5  2021-9-24 torch 1.9.1 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 11264.0MB)

Fusing layers... 
Model Summary: 224 layers, 7266973 parameters, 0 gradients
Adding AutoShape... 
image 1/1: 720x1280 2 persons, 2 ties
Speed: 15.6ms pre-process, 15.6ms inference, 0.0ms NMS per image at shape (1, 3, 384, 640)

Process finished with exit code 0

So I run twice for detect.py it skipped the preprocess?
Speed: 0.0ms pre-process, 15.6ms inference, 0.0ms NMS per image at shape (1, 3, 640, 640)
And hub didn't?
Speed: 15.6ms pre-process, 15.6ms inference, 0.0ms NMS per image at shape (1, 3, 384, 640)

**Comment by glenn-jocher** [comment]
Hub pre-process times include all ops required for dataloading, whereas detect.py employs a seperate dataloader that is not included in profiling.
